### ISSUE TITLE
Disable heartbeats in BlockingConnection

### ISSUE THREAD

**Issue by carlhoerberg** [pika/pika#636]
As BlockingConnection doesn't handle heartbeats in the background and the heartbeat_interval can't override the servers suggested heartbeat interval (that's a bug too), i suggest that heartbeats should be disabled by default (rely on TCP keep-alive instead). 

If processing a task in a consume block takes longer time then the server suggested heartbeat interval, the connection will be closed by the server and the client won't be able to ack the message when it's done processing. 

The next best thing you can do is to allow to use a larger heartbeat interval then the server suggested interval (all other clients allow that and the server will accept it). The third best thing would be to document this behaviour. 

Handling heartbeats in a separate thread would of course be the best thing but I guess that goes against what the BlockingAdapter is all about..

**Comment by prawn-cake** [comment]
+1
still reproduced

**Comment by vitaly-krugl** [comment]
I agree that it would be great if BlockingConnection could handle heartbeat timeouts in the background. This issue has become more noticeable since a recent version of RabbitMQ broker reduced its default heartbeat timeout to 1 minute from almost 10 minutes.

Short of that, I think it would be good to have a way in pika so that a user could ask for a bigger heartbeat timeout and have it respected.

**Comment by vitaly-krugl** [comment]
Michael Klishin (Pivotal/RabbitMQ) recommends that the clients should follow what the Java client does and accept the max of the two values, not the min. See https://groups.google.com/forum/#!topic/rabbitmq-users/FpUdTaIr8CI.

I opened #665 for picking the max during heartbeat timeout negotiation.

**Comment by nam4dev** [comment]
I've actually found a workaround, as I bumped into that bug as well.

**This is only a workaround**
* I initialize a new connection each time I need to produce (instead of keeping connection alive).
* I then, close the connection when data is produced.

I did not reproduce the bug since then!

Hope it helps :)

**Comment by prawn-cake** [comment]
Not production-ready workaround, if you need to stay performant you have to reuse your connection --> heartbeats are required

**Comment by vitaly-krugl** [comment]
Once PR #666 is merged, you should at least be able to configure a longer timeout.

Please note that prior to RabbitMQ v3.5.5, the default heartbeat timeout from RabbitMQ was 580 seconds (almost 10 minutes).  RabbitMQ v3.5.5 dropped the default to just 60 seconds. See https://www.rabbitmq.com/heartbeats.html

**Comment by nam4dev** [comment]
@prawn-cake 

Depends on your design and needs indeed :)
I agree with you **heartbeats are required**!!

But just a thought, 
* AMQP is just a protocol defined over TCP, right?
* HTTP is a protocol over TCP too, right?

Therefore, when you use an HTTP client, say for instance `requests` do you really need to keep your connection alive when sending a POST request ?
I mean, in term of `Producer`, is that such a big deal to open a connection, send data on the wire, then close?

Even in production for some cases at least, I think this is an acceptable **workaround**, though, I agree not the best!

**Comment by vitaly-krugl** [comment]
Regarding the workaround, it depends: if you have to publish many messages at once, then opening and closing an AMQP connection for every message may punish your performance (and broker's as well?). Keep in mind that setting up an AMQP connection involves a TCP/IP connection + AMQP handshake; and then opening a channel is another handshake; all these operations add latencies.

There are other scenarios, where opening a connection per messages would be entriely reasonable.

**Comment by nam4dev** [comment]
Ok thx!

**Comment by prawn-cake** [comment]
Indeed in most of cases keeping single connection as long as possible better than open/close it - it's not reasonable, but affordable in some cases.
Exceptions can be if connections are limited (hardware or server limitations) or you wanna to hide stuff - shady business.

@nam4dev this is reasonable and for *HTTP* as well (hereby I mean rich web apps mostly) - keep as less connections as possible and as long as possible and this is one of the drawbacks for HTTP 1/X with its workarounds with "Keep-Alive" *HTTP* headers, and this is one of the bullet points of *HTTP 2*, cuz it obviously increasing performance and getting rid of this handshake hell.

So yes, we have to have working heartbeats =)

**Comment by vitaly-krugl** [comment]
@nickcash's PR #666 "Pick the max of the heartbeat timeouts as result of negotiation" has just been merged into pika master. We're going to keep the current default behavior (accept server's value), but an app can now configure the heartbeat timeout for a given connection via ConnectionParameters or URLParameters as follows:
- the tuning algorithm will pick the max(client_value, server_value); 0 is treated as the highest possible value  (zero=disable), and will disable heartbeats altogether for the given connection.

**Comment by vitaly-krugl** [comment]
Closing issue, since using pika master, an application can now set a larger heartbeat timeout per PR #666. Thank you.
